# 项目：评估和清理英国电商公司销售数据

## 分析目标

此数据分析的目的是，根据市场销售数据，挖掘畅销产品，以便制定更有效的市场策略来提升营收。

本实战项目的目的在于练习评估数据干净和整洁度，并且基于评估结果，对数据进行清洗，从而得到可供下一步分析的数据。

## 简介

原始数据集记录了一家英国在线零售公司在2010年12月1日至2011年12月9日期间的所有交易情况，涵盖了该公司在全球不同国家和地区的业务数据。该公司主要销售覆盖各个场景的礼品，包括但不限于生日礼品、结婚纪念品、圣诞礼品等等。该公司的客户群体主要包括批发商和个人消费者，其中批发商占据了相当大的比例。

数据每列的含义如下：
- `InvoiceNo`: 发票号码。6位数，作为交易的唯一标识符。如果这个代码以字母“c”开头，表示这笔交易被取消。
- `StockCode`: 产品代码。5位数，作为产品的唯一标识符。
- `Description`: 产品名称。
- `Quantity`: 产品在交易中的数量。
- `InvoiceDate`: 发票日期和时间。交易发生的日期和时间。
- `UnitPrice`: 单价。价格单位为英镑（£）。
- `CustomerID`: 客户编号。5位数，作为客户的唯一标识符。
- `Country`: 国家名称。客户所居住的国家的名称。

## 读取数据

导入数据分析所需要的库，并通过Pandas的`read_csv`函数，将原始数据文件"e_commerce.csv"里的数据内容，解析为DataFrame，并赋值给变量`original_data`。

In [1]:
import pandas as pd

In [2]:
original_data=pd.read_csv('e_commerce.csv')

In [3]:
original_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

## 评估数据

在这一部分，我将对在上一部分建立的`original_data`这个DataFrame所包含的数据进行评估。
数据的评估主要从两方面进行：数据结构和数据内容，即整齐度和干净度。
数据的结构性问题指不符合“每列是一个变量，每行是一个观察值，每个单元格是一个观察值”这三个标准。
数据的内容性问题指数据缺失，重复数据，无效数据等。

### 评估数据整齐度（数据结构评估）

In [4]:
original_data.sample(10)

InvoiceNo StockCode                     Description  Quantity  \
248552    558861     21471    STRAWBERRY RAFFIA FOOD COVER         6   
156913    550183     21754        HOME BUILDING BLOCK WORD         9   
10421     537237     22107              PIZZA PLATE IN BOX         4   
342747    566913     23355      HOT WATER BOTTLE KEEP CALM         2   
214618    555591     22677   FRENCH BLUE METAL DOOR SIGN 2        10   
77068     542692     22570        FELTCRAFT CUSHION RABBIT         3   
538793    581414     21408       SPOTTY PINK DUCK DOORSTOP         1   
376322    569521     22141  CHRISTMAS CRAFT TREE TOP ANGEL         2   
321591    565152     22208        WOOD STAMP SET THANK YOU         1   
92666     544202     20685           DOORMAT RED RETROSPOT         1   

            InvoiceDate  UnitPrice  CustomerID         Country  
248552   7/4/2011 12:19       3.75     13373.0  United Kingdom  
156913  4/14/2011 17:58       5.95     14105.0  United Kingdom  
10421    12/6/2010 9:58       7.62         NaN  United Kingdom  
342747  9/15/2011 14:07       4.95     17602.0  United Kingdom  
214618    6/6/2011 9:59       1.25     15073.0  United Kingdom  
77068   1/31/2011 12:35       3.75     14467.0  United Kingdom  
538793  12/8/2011 14:39       1.95     14730.0  United Kingdom  
376322  10/4/2011 14:33       2.10     16474.0  United Kingdom  
321591   9/1/2011 14:42       0.83     17799.0  United Kingdom  
92666   2/17/2011 10:29      16.63         NaN  United Kingdom

从抽样的10行数据来看，数据符合“每列是一个变量，每行是一个观察值，每个单元格是一个值”具体来看每行是关于某件商品的一次交易，每列是商品交易相关的各个变量，因此不存在数据结构性问题

### 评估数据干净度（数据内容评估）

In [5]:
original_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


从输出结果来看，数据总共有541909条观察值，而Description和CustomerID变量存在缺失值。

此外，**InvoiceDate数据的类型应该为日期，CustomerID的数据类型应该为字符串，应当进行数据格式转换**。

#### 评估缺失数据

在了解到Description变量存在缺失值后，根据缺失条件，提取缺失观察值。

In [7]:
original_data[original_data['Description'].isnull()]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
622       536414     22139         NaN        56  12/1/2010 11:52        0.0   
1970      536545     21134         NaN         1  12/1/2010 14:32        0.0   
1971      536546     22145         NaN         1  12/1/2010 14:33        0.0   
1972      536547     37509         NaN         1  12/1/2010 14:33        0.0   
1987      536549    85226A         NaN         1  12/1/2010 14:34        0.0   
...          ...       ...         ...       ...              ...        ...   
535322    581199     84581         NaN        -2  12/7/2011 18:26        0.0   
535326    581203     23406         NaN        15  12/7/2011 18:31        0.0   
535332    581209     21620         NaN         6  12/7/2011 18:35        0.0   
536981    581234     72817         NaN        27  12/8/2011 10:33        0.0   
538554    581408     85175         NaN        20  12/8/2011 14:06        0.0   

        CustomerID         Country  
622            NaN  United Kingdom  
1970           NaN  United Kingdom  
1971           NaN  United Kingdom  
1972           NaN  United Kingdom  
1987           NaN  United Kingdom  
...            ...             ...  
535322         NaN  United Kingdom  
535326         NaN  United Kingdom  
535332         NaN  United Kingdom  
536981         NaN  United Kingdom  
538554         NaN  United Kingdom  

[1454 rows x 8 columns]

总共有1454条Description缺失的观察值。仅仅从输出结果来看，缺失Description的交易记录，UnitPrice都为0，为了验证猜想，我们增加筛选的条件，查看是否存在，Description缺失同时UnitPrice不为0的数据。

In [8]:
original_data[original_data['Description'].isnull() & original_data['UnitPrice']!=0 ]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

筛选出来的数据的数量为0条，说明缺失Description的数据，UnitPrice都为0，即不具备有效的UnitPrice值。

Description表示产品名称，UnitPrice表示产品单价，都是后续进行商品交易分析的重要变量，如果它们同时存在缺失或无效情况，我们认为数据无法提供有效的含义，**因此可以考虑删除Description缺失的数据**。

CustomerID变量也存在缺失值，同样根据条件，提取CustomerID缺失的观察值

In [9]:
original_data[original_data['CustomerID'].isnull()]

InvoiceNo StockCode                      Description  Quantity  \
622       536414     22139                              NaN        56   
1443      536544     21773  DECORATIVE ROSE BATHROOM BOTTLE         1   
1444      536544     21774  DECORATIVE CATS BATHROOM BOTTLE         2   
1445      536544     21786               POLKADOT RAIN HAT          4   
1446      536544     21787            RAIN PONCHO RETROSPOT         2   
...          ...       ...                              ...       ...   
541536    581498    85099B          JUMBO BAG RED RETROSPOT         5   
541537    581498    85099C   JUMBO  BAG BAROQUE BLACK WHITE         4   
541538    581498     85150    LADIES & GENTLEMEN METAL SIGN         1   
541539    581498     85174                S/4 CACTI CANDLES         1   
541540    581498       DOT                   DOTCOM POSTAGE         1   

            InvoiceDate  UnitPrice  CustomerID         Country  
622     12/1/2010 11:52       0.00         NaN  United Kingdom  
1443    12/1/2010 14:32       2.51         NaN  United Kingdom  
1444    12/1/2010 14:32       2.51         NaN  United Kingdom  
1445    12/1/2010 14:32       0.85         NaN  United Kingdom  
1446    12/1/2010 14:32       1.66         NaN  United Kingdom  
...                 ...        ...         ...             ...  
541536  12/9/2011 10:26       4.13         NaN  United Kingdom  
541537  12/9/2011 10:26       4.13         NaN  United Kingdom  
541538  12/9/2011 10:26       4.96         NaN  United Kingdom  
541539  12/9/2011 10:26      10.79         NaN  United Kingdom  
541540  12/9/2011 10:26    1714.17         NaN  United Kingdom  

[135080 rows x 8 columns]

CustomerID为客户编号，用于唯一区分客户，不是后续商品交易分析的必要变量，并且从输出结果来看，一些CustomerID缺失的交易数据任然有效，因此保留CustomerID缺失的观察值

#### 评估重复数据

根据数据变量的含义来看，虽然InvoiceNo、StockCode和CustomerID都是唯一标识符，但是一次交易中，可能包含多件商品，因此InvoiceNo可以存在重复值。不同交易可以包含同件商品，因此StockCode可以存在重复值。一个客户可以进行多次交易或下单多个商品，因此CustomerID可以存在重复值。

#### 评估不一致数据

不一致数据有可能存在于Country变量中，需要查看是否存在多个值指代同一个国家的情况。

In [10]:
original_data['Country'].value_counts()

Country
United Kingdom          495266
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
China                      288
Singapore                  229
USA                        218
UK                         211
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United States               73
United Arab Emirates        68


**从Country变量值来看，"USA"、"United States"均在表示美国，"United Kingdom"、"UK"、"U.K."均在表示英国，因此应该在清洗步骤对这些值进行统一，只保留一个指代值**。

#### 评估无效或错误数据

可以调用dataframe的describe方法，对数值的统计信息进行快速了解

In [11]:
original_data.describe()

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

从输出结果来看，Quantity和UnitPrice存在负数，会对后续数值分析产生影响。

因此，我们先筛选出Quantity数值为负数的观察值，进一步评估其含义。

In [12]:
original_data[original_data['Quantity']<0]

InvoiceNo StockCode                       Description  Quantity  \
141      C536379         D                          Discount        -1   
154      C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235      C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236      C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237      C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   
...          ...       ...                               ...       ...   
540449   C581490     23144   ZINC T-LIGHT HOLDER STARS SMALL       -11   
541541   C581499         M                            Manual        -1   
541715   C581568     21258        VICTORIAN SEWING BOX LARGE        -5   
541716   C581569     84978  HANGING HEART JAR T-LIGHT HOLDER        -1   
541717   C581569     20979     36 PENCILS TUBE RED RETROSPOT        -5   

            InvoiceDate  UnitPrice  CustomerID         Country  
141      12/1/2010 9:41      27.50     14527.0  United Kingdom  
154      12/1/2010 9:49       4.65     15311.0  United Kingdom  
235     12/1/2010 10:24       1.65     17548.0  United Kingdom  
236     12/1/2010 10:24       0.29     17548.0  United Kingdom  
237     12/1/2010 10:24       0.29     17548.0  United Kingdom  
...                 ...        ...         ...             ...  
540449   12/9/2011 9:57       0.83     14397.0  United Kingdom  
541541  12/9/2011 10:28     224.69     15498.0  United Kingdom  
541715  12/9/2011 11:57      10.95     15311.0  United Kingdom  
541716  12/9/2011 11:58       1.25     17315.0  United Kingdom  
541717  12/9/2011 11:58       1.25     17315.0  United Kingdom  

[10624 rows x 8 columns]

从筛选结果来看，`Quantity`变量为负数的观察值，`InvoiceNo`似乎均以C开头，表示订单被取消。

但此猜测不一定准确，为了验证猜想，我们增加筛选条件，看是否存在`Quantity`变量为负数且`InvoiceNo`不以C开头的观察值。

In [13]:
original_data[(original_data['Quantity']<0) & (original_data['InvoiceNo'].str[0]!='C')]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
2406      536589     21777         NaN       -10  12/1/2010 16:50        0.0   
4347      536764    84952C         NaN       -38  12/2/2010 14:42        0.0   
7188      536996     22712         NaN       -20  12/3/2010 15:30        0.0   
7189      536997     22028         NaN       -20  12/3/2010 15:30        0.0   
7190      536998     85067         NaN        -6  12/3/2010 15:30        0.0   
...          ...       ...         ...       ...              ...        ...   
535333    581210     23395       check       -26  12/7/2011 18:36        0.0   
535335    581212     22578        lost     -1050  12/7/2011 18:38        0.0   
535336    581213     22576       check       -30  12/7/2011 18:38        0.0   
536908    581226     23090     missing      -338   12/8/2011 9:56        0.0   
538919    581422     23169     smashed      -235  12/8/2011 15:24        0.0   

        CustomerID         Country  
2406           NaN  United Kingdom  
4347           NaN  United Kingdom  
7188           NaN  United Kingdom  
7189           NaN  United Kingdom  
7190           NaN  United Kingdom  
...            ...             ...  
535333         NaN  United Kingdom  
535335         NaN  United Kingdom  
535336         NaN  United Kingdom  
536908         NaN  United Kingdom  
538919         NaN  United Kingdom  

[1336 rows x 8 columns]

以上猜想被证实错误，因为还存在Quantity变量为负数且InvoiceNo不以C开头的观察值。

但根据以上输出结果，这些筛选出的观察值的UnitPrice观察值均为0，因此增加UnitPrice的条件进行验证。

In [14]:
original_data[(original_data["Quantity"] < 0)&(original_data["InvoiceNo"].str[0] != "C")&(original_data["UnitPrice"] != 0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

根据输出结果，猜想得到验证，当Quantity变量为负数的时候，观察值满足以下条件之一：
- InvoiceNo以C开头，表示订单被取消
- UnitPrice为0，说明单价为0英镑

这些交易数据均不是有效成交数据，不贡献销售，不在后续分析范围内，因此我们将在数据清理步骤，**将`Quantity`为负数的观察值删除**。

接下来筛选UnitPrice为负数的观察值

In [16]:
original_data[original_data["UnitPrice"] < 0]

InvoiceNo StockCode      Description  Quantity      InvoiceDate  \
299983   A563186         B  Adjust bad debt         1  8/12/2011 14:51   
299984   A563187         B  Adjust bad debt         1  8/12/2011 14:52   

        UnitPrice  CustomerID         Country  
299983  -11062.06         NaN  United Kingdom  
299984  -11062.06         NaN  United Kingdom

从输出结果来看，UnitPrice为负数的观察值都是坏账调账，不属于实际商品交易数据，**因此也在数据清理步骤中也将UnitPrice为负数的观察值删除**。

## 清理数据

根据前面评估部分得到的结论,我们需要执行以下数据清理：
- 把`InvoiceDate`变量的数据类型转换为为日期时间类型
- 把`CustomerID`变量的数据类型转化为字符串
- 把`Description`变量缺失的观察值删除
- 把`Country`中的变量值`"USA"`更改为`"United states"`
- 把`Country`中的变量值`"UK"`,`"U.K."`更改为`"United Kingdom"`
- 把`Quantity`变量值为负数的观察值删除
- 把`UnitPrice`变量值为负数的观察值删除

为了区分经过清洗的数据和原始的数据，我们创建新的变量cleaned_data, 让它为original_data复制出的副本，之后清洗的步骤都将运用到cleaned_data上。

In [17]:
cleaned_data=original_data.copy()
cleaned_data.head(10)

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
5    536365     22752         SET 7 BABUSHKA NESTING BOXES         2   
6    536365     21730    GLASS STAR FROSTED T-LIGHT HOLDER         6   
7    536366     22633               HAND WARMER UNION JACK         6   
8    536366     22632            HAND WARMER RED POLKA DOT         6   
9    536367     84879        ASSORTED COLOUR BIRD ORNAMENT        32   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom  
5  12/1/2010 8:26       7.65     17850.0  United Kingdom  
6  12/1/2010 8:26       4.25     17850.0  United Kingdom  
7  12/1/2010 8:28       1.85     17850.0  United Kingdom  
8  12/1/2010 8:28       1.85     17850.0  United Kingdom  
9  12/1/2010 8:34       1.69     13047.0  United Kingdom

把`InvoiceDate`变量的数据类型转换为为日期时间类型：

In [18]:
cleaned_data['InvoiceDate']=pd.to_datetime(cleaned_data['InvoiceDate'])
cleaned_data['InvoiceDate']

0        2010-12-01 08:26:00
1        2010-12-01 08:26:00
2        2010-12-01 08:26:00
3        2010-12-01 08:26:00
4        2010-12-01 08:26:00
                 ...        
541904   2011-12-09 12:50:00
541905   2011-12-09 12:50:00
541906   2011-12-09 12:50:00
541907   2011-12-09 12:50:00
541908   2011-12-09 12:50:00
Name: InvoiceDate, Length: 541909, dtype: datetime64[ns]

把`CustomerID`变量的数据类型转化为字符串:

In [19]:
cleaned_data['CustomerID']=cleaned_data['CustomerID'].astype(str)
cleaned_data['CustomerID']

0         17850.0
1         17850.0
2         17850.0
3         17850.0
4         17850.0
           ...   
541904    12680.0
541905    12680.0
541906    12680.0
541907    12680.0
541908    12680.0
Name: CustomerID, Length: 541909, dtype: object

把CustomerID变量值结尾的".0"删除：

In [20]:
cleaned_data['CustomerID']=cleaned_data['CustomerID'].str.slice(0,-2)
cleaned_data['CustomerID']

0         17850
1         17850
2         17850
3         17850
4         17850
          ...  
541904    12680
541905    12680
541906    12680
541907    12680
541908    12680
Name: CustomerID, Length: 541909, dtype: object

把`Description`变量缺失的观察值删除, 并查看删除后该列的空缺值个数总和

In [22]:
cleaned_data.dropna(subset=['Description'],inplace=True)

In [23]:
cleaned_data['Description'].isnull().sum()

0

把`Country`变量值`"USA"`替换为`"United States"`，并检查替换后`"USA"`变量值个数：

In [24]:
cleaned_data['Country']=cleaned_data['Country'].replace({'USA':'United States'})

In [25]:
len(cleaned_data[cleaned_data["Country"] == "USA"])

0

把Country变量值"UK"、"U.K."替换为"United Kingdom"，并检查替换后"UK"和"U.K."变量值个数：

In [27]:
cleaned_data["Country"] = cleaned_data["Country"].replace({"UK": "United Kingdom", "U.K.": "United Kingdom"})

In [28]:
print(len(cleaned_data[cleaned_data["Country"] == 'UK'] ))
print(len(cleaned_data[cleaned_data["Country"] == 'U.K.'] ))

0
0


把`Quantity`变量值为负数的观察值删除，并检查替换后`Quantity`变量值为负数的个数：

In [29]:
cleaned_data=cleaned_data[cleaned_data['Quantity']>=0]

In [30]:
len(cleaned_data[cleaned_data['Quantity']<0])

0

把`UnitPrice`变量值为负数的观察值删除，并检查替换后`UnitPrice`变量值为负数的个数：

In [31]:
cleaned_data=cleaned_data[cleaned_data['UnitPrice']>=0]

In [32]:
len(cleaned_data[cleaned_data['UnitPrice']<0])

0

## 保存清理后的数据

In [35]:
cleaned_data.head(10)

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
5    536365     22752         SET 7 BABUSHKA NESTING BOXES         2   
6    536365     21730    GLASS STAR FROSTED T-LIGHT HOLDER         6   
7    536366     22633               HAND WARMER UNION JACK         6   
8    536366     22632            HAND WARMER RED POLKA DOT         6   
9    536367     84879        ASSORTED COLOUR BIRD ORNAMENT        32   

          InvoiceDate  UnitPrice CustomerID         Country  
0 2010-12-01 08:26:00       2.55      17850  United Kingdom  
1 2010-12-01 08:26:00       3.39      17850  United Kingdom  
2 2010-12-01 08:26:00       2.75      17850  United Kingdom  
3 2010-12-01 08:26:00       3.39      17850  United Kingdom  
4 2010-12-01 08:26:00       3.39      17850  United Kingdom  
5 2010-12-01 08:26:00       7.65      17850  United Kingdom  
6 2010-12-01 08:26:00       4.25      17850  United Kingdom  
7 2010-12-01 08:28:00       1.85      17850  United Kingdom  
8 2010-12-01 08:28:00       1.85      17850  United Kingdom  
9 2010-12-01 08:34:00       1.69      13047  United Kingdom

完成数据清理后，把干净整齐的数据保存到新的文件里，文件名为e_commerce_cleaned.csv。

In [33]:
cleaned_data.to_csv('e_commerce_cleaned.csv',index=False)

In [36]:
pd.read_csv('e_commerce_cleaned.csv').head()

C:\Users\HP\AppData\Local\Temp\ipykernel_15548\563775235.py:1: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.
  pd.read_csv('e_commerce_cleaned.csv').head()


InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

           InvoiceDate  UnitPrice CustomerID         Country  
0  2010-12-01 08:26:00       2.55      17850  United Kingdom  
1  2010-12-01 08:26:00       3.39      17850  United Kingdom  
2  2010-12-01 08:26:00       2.75      17850  United Kingdom  
3  2010-12-01 08:26:00       3.39      17850  United Kingdom  
4  2010-12-01 08:26:00       3.39      17850  United Kingdom